### PR TITLE
#5791: Address errors found by lgtm.com

### PR DESCRIPTION
**Pull request by sturmer**
### CATEGORY

Choose one

- [ ] Bug Fix
- [x] Enhancement (new features, refinement)
- [ ] Refactor
- [x] Add tests
- [ ] Build / Development Environment
- [ ] Documentation

### SUMMARY
The website LGTM.com has detected a few errors from static analysis. This patch contains a possible fix for the first of them ("URL redirection from remote source") that follows the website's own recommendation: Check the input against some known source before using it. I've checked that the dashboard ID provided actually exists in DB and if it doesn't I return a 404.
I plan to solve the other 4 issues labeled as "Error" but wanted to get some early feedback before moving on.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
Not applicable.

### TEST PLAN
I've written a unit test that can be run e.g. like:
`tox -e py37 -- tests/access_tests.py:RequestAccessTests.test_dashboard_endpoint_malicious_redirect`

### ADDITIONAL INFORMATION

- [x] Has associated issue: #5791 
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### REVIEWERS
Please @mistercrunch as original reporter but feel free to re-assign as you see fit.

### CODE REVIEW

File: superset/connectors/druid/models.py [hunk 1]
```diff
@@ -83,6 +84,8 @@ def __init__(self, name, field_names, function):
 class CustomPostAggregator(Postaggregator):
     """A way to allow users to specify completely custom PostAggregators"""
     def __init__(self, name, post_aggregator):
+        # TODO(gianluca): Is this the right way? should I use the arg post_aggregator?
+        Postaggregator.__init__(self, None, None, name)
```
Comment (john-bodley) [inline superset/connectors/druid/models.py]: Please use `super` instead.

File: superset/stats_logger.py [hunk 2]
```diff
@@ -82,6 +82,7 @@ def __init__(self, host='localhost', port=8125,
             If statsd_client argument is given, all other arguments are ignored and the
             supplied client will be used to emit metrics.
             """
+            BaseStatsLogger.__init__(self, prefix)
```
Comment (john-bodley) [inline superset/stats_logger.py]: Please use `super` instead.

File: superset/utils/core.py [hunk 3]
```diff
@@ -126,9 +126,13 @@ def __repr__(self):
...
         return self.func.__doc__
 
+    def __copy__(self):
+        return _memoized(self.func, self.watch)
+
     def __get__(self, obj, objtype):
         if not self.is_method:
-            self.is_method = True
+            copy = self.__copy__()
+            return functools.partial(copy.__call__, obj)
```
Comment (john-bodley) [inline superset/utils/core.py]: Could you provide some more context as to what you’re trying to achieve here?

File: tests/access_tests.py [hunk 4]
```diff
@@ -428,6 +428,14 @@ def test_approve(self, mock_send_mime):
             session.delete(security_manager.find_role(TEST_ROLE_NAME))
             session.commit()
 
+    def test_dashboard_endpoint_malicious_redirect(self):
+        # todo(gianluca.ciccarelli@bolt.eu): I need a dashboard ID that will
```
Comment (john-bodley) [inline tests/access_tests.py]: You could try using `-1` (though I’m unsure if that works), otherwise maybe querying for the maximum dashboard ID and adding one, or using `9999`.

File: superset/connectors/druid/models.py [hunk 5]
```diff
@@ -71,6 +71,7 @@ def _fetch_metadata_for(datasource):
 
 class JavascriptPostAggregator(Postaggregator):
     def __init__(self, name, field_names, function):
+        Postaggregator.__init__(self, function, field_names, name)
```
Comment (mistercrunch) [inline superset/connectors/druid/models.py]: Edits to this file seems unrelated to your PR title

Comment (sturmer) [reply superset/connectors/druid/models.py]: I agree. The thing is that I started on the issue but gave a title related to the first I had started tackling. Do you suggest changing the title or split the PR into a few ones? (I'd favor the first.)

File: superset/utils/core.py (hunk 3 above)
Comment (sturmer) [reply superset/utils/core.py]: The analyzer at lgtm.com finds this assignment harmful (https://lgtm.com/rules/910074/). 

The problem is (AFAIU) that we are mutating a descriptor object, and this can cause surprises to different instances of the same class. The recommendation is to create a copy that contains the desired state and return that instead.

In a patch I haven't pushed yet, I do a `copy.is_method = True` before returning as I do here. I wanted to copy the present instance, mutate it, and return it.

This, however, makes the unit tests `test_memoized_on_methods_with_watches` and `test_memoized_on_methods` fail.

Comment (sturmer) [reply superset/utils/core.py]: On further thought I think that the code uses this pattern to actually achieve the desired result. What I propose, in this case, is to either silence the alert with `# lgtm[py/mutable-descriptor]`, or to just ignore it. Maybe a comment in the code would be enough?

Comment (sturmer) [reply superset/utils/core.py]: @john-bodley kindly when you find a minute can you advise on this?

**Comment by sturmer** [comment]
@mistercrunch @john-bodley please when you have time can you have a look at this? In the meanwhile, I'll start addressing the lower-priority alerts (I remember seeing many unused imports, they should be safe to remove).

**Comment by sturmer** [comment]
To be clear, I think this patch could be merged as-is, but I'll add some lower-priority work to it that won't hurt. At least that's the plan 😄

**Comment by sturmer** [comment]
I need to re-scope this one in light of the more recent alerts. I think that the important is that I manage to fix the 4 errors currently pointed out.

**Comment by sturmer** [comment]
Now it should fix the 4 Errors and tackle a few Warnings as a bonus. It would be useful to point LGTM.com to this branch and check that that is actually the case, but I haven't been able to.

Is there still interest in going through with this? @john-bodley @mistercrunch
